lib/index.js

```diff
@@ -0,0 +1,532 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+// The top-level APIs offered by this library are `ClientConnection` and
+// `ServerConnection` classes.  They each take a callback to be used for
+// sending data to the remote peer, and operate like this:
+//
+//    conn = new ClientConnection(async function send_data_to_server(data) {
+//      // application-specific sending logic here.
+//    })
+//
+//    await conn.initialize(psk, psk_id)
+//
+//    // Send data to the server by calling `send`,
+//    // which will use the callback provided in the constructor.
+//    // A single `send()` by the application may result in multiple
+//    // invokations of the callback.
+//
+//    await conn.send('application-level data')
+//
+//    // When data is received from the server, push it into
+//    // the connection and let it return any decrypted app-level data.
+//    // There might not be any app-level data if it was protocol control message,
+//    // and the receipt of the data might trigger additional calls to the
+//    // send callback for protocol control purposes.
+//
+//    serverSocket.on('data', async encrypted_data => {
+//      const plaintext = await conn.recv(data)
+//      if (plaintext !== null) {
+//        do_something_with_app_level_data(plaintext)
+//      }
+//    })
+//
+//    // It's good practice to explicitly close the connection
+//    // when finished.  This will send a "closed" notification
+//    // to the server.
+//
+//    await conn.close()
+//
+// The `ServerConnection` API operates similarly; the distinction is mainly
+// in which side is expected to send vs receieve during the protocol handshake.
+
+// !!!!!!!
+// !!
+// !!   N.B. We have not yet implemented the actual encryption bits!
+// !!
+// !!!!!!!
+//
+// This first version of the code uses the same "framework" as we would use for
+// implementing TLS1.3, including the basic sequence of messages between client and
+// server, and the use of binary encoding in an ArrayBuffer for data processing.
+// Using an implementation with proper crypto should feel identical to using
+// this mock version, except it won't have "Insecure" in the class name...
+
+class InsecureConnection {
+    constructor(send_callback) {
+        this.state = new State_UNINITIALIZED(this)
+        this.pending_send_data = []
+        this.incoming_buffer = new IncomingRecordBuffer()
+        this.outgoing_buffer = new OutgoingRecordBuffer()
+        this.send_callback = send_callback
+        // We use these for serializing calls to `send()` and `recv()`.
+        this.send_queue = Promise.resolve()
+        this.recv_queue = Promise.resolve()
+    }
+
+    async initialize(psk, psk_id) {
+        this.psk = coerceToBytes(psk) // XXX TODO: probably the app should pass this in as a Uint8Array.
```

Comment:
We should make it so, good call.

Reply:
Would it be useful to leverage the type system here so we don't call initialize twice? e.g. make `initialize` static and return an instance or make it return another class instance?
It might be too much of a refactor so I'm also happy with `if initialized then throw boom` :)

Reply:
I am a big fan of using the type system for things, let's give it a shot...

The only reason these aren't arguments to the constructor is because this function wants to be `async`, so a static async initializer probably makes sense.